plinth/modules/first_boot/middleware.py

```diff
@@ -37,19 +38,55 @@ class FirstBootMiddleware(object):
     def process_request(request):
         """Handle a request as Django middleware request handler."""
         state = kvstore.get_default('firstboot_state', 0)
-
-        firstboot_index_url = reverse('first_boot:index')
-        user_requests_firstboot = request.path.startswith(firstboot_index_url)
-
+        user_requests_firstboot = is_firstboot(request.path)
+        user_requests_login = request.path.startswith(reverse(settings.LOGIN_URL))
         help_index_url = reverse('help:index')
         user_requests_help = request.path.startswith(help_index_url)
+        if not user_requests_login and not user_requests_help:
+            if state == 1 and user_requests_firstboot:
+                return HttpResponseRedirect(reverse('index'))
+            elif state == 0 and not user_requests_firstboot:
+                url = next_step()
+                return HttpResponseRedirect(reverse(url))
+
+
+def is_firstboot(path):
+    """
+    Returns whether the path is a firstboot step url
+    :param path: path of current url
+    :return: true if its a first boot url false otherwise
+    """
+    steps = get_firstboot_steps()
+    for step in steps:
+        if reverse(step.get('url')) == path:
+            return True
+    return False
+
+
+def get_firstboot_steps():
+    """Returns all firstboot steps"""
+    steps = []
+    modules = module_loader.loaded_modules
+    for (module_name, module_object) in modules.items():
+        if getattr(module_object, 'first_boot_steps', None):
+            for step in module_object.first_boot_steps:
+                steps.append(step)
+    steps = sorted(steps, key=itemgetter('order'))
+    return steps
+
+
+def next_step():
+    """ Returns the next first boot step required to run """
+    steps = get_firstboot_steps()
```

Comment:
Would it be possible to keep the results of get_firstboot_steps(), the first time it is called, so that we don't have to run it for every step?